### PR TITLE
feat(marketplace) add as a new pallet extending acurast core

### PR DESCRIPTION
**Pull request by gitsimon**

### CODE REVIEW

File: pallets/marketplace/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+
+    use crate::traits::*;
+    use crate::types::*;
+    use crate::utils::*;
+
+    #[pallet::config]
+    pub trait Config:
+        frame_system::Config
+        + pallet_timestamp::Config
+        + pallet_assets::Config
```
Comment (godenzim) [inline pallets/marketplace/src/lib.rs:46]: @gitsimon these 2 are not needed and can be removed

File: pallets/marketplace/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+            total
+                .with_amount(
+                    reward_amount
+                        .checked_mul(&((extra.slots as u128).into()))
+                        .ok_or(Error::<T>::RewardCalculationOverflow)?
+                        .into(),
+                )
+                .map_err(|_| Error::<T>::RewardConversionFailed)?;
+            T::RewardManager::lock_reward(total.clone(), T::Lookup::unlookup(who.clone()))
+                .map_err(|_| Error::<T>::InvalidPayment)?;
```
Comment (godenzim) [inline pallets/marketplace/src/lib.rs:342]: @gitsimon `T::RewardManager::lock_reward` is currently still called in the register extrinsic in `pallet_acurast` before the `register_hook` is called. So it is being called twice now.

File: pallets/marketplace/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+            ensure!(
+                job_status != JobStatus::Assigned,
+                Error::<T>::CannotFulfillJobWhenNotAssigned
+            );
+
+            T::RewardManager::pay_reward(
+                job_registration.reward.clone(),
+                T::Lookup::unlookup(who.clone()),
+            )
+            .map_err(|_| Error::<T>::FailedToPay)?;
```
Comment (godenzim) [inline pallets/marketplace/src/lib.rs:412]: @gitsimon same here `T::RewardManager::pay_reward` is called already in the `fulfill` extrinsic in `pallet_acurast` before the `fulfill_hook` is called. So now it is called twice.

**Comment by godenzim** [comment]
@gitsimon please run `cargo fmt` to make sure the code is properly formatted if your IDE is not doing it for you

File: pallets/acurast/src/lib.rs [hunk 4]
```diff
@@ -10,7 +10,7 @@ mod benchmarking;
 
 mod attestation;
 pub mod payments;
-mod traits;
+pub mod traits;
```
Comment (godenzim) [inline pallets/acurast/src/lib.rs:13]: @gitsimon why did you need to make it `pub`? everything is re-exposted publicly on line 21

File: pallets/marketplace/src/lib.rs (hunk 1 above)
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs]: all warnings such as unused imports are fixed

File: pallets/marketplace/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+        /// Cannot get the certificate ID.
+        CannotGetCertificateId,
+        /// Failed to convert the attestation to its bounded type.
+        AttestationToBoundedTypeConversionFailed,
+        /// Timestamp error.
+        FailedTimestampConversion,
+        /// Certificate was revoked.
+        RevokedCertificate,
+        /// Origin is not allowed to update the certificate revocation list.
+        CertificateRevocationListUpdateNotAllowed,
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs]: removed

File: pallets/acurast/src/lib.rs [hunk 6]
```diff
@@ -315,43 +317,34 @@ pub mod pallet {
...
-                T::Lookup::unlookup(who.clone()),
+            <T as Config>::JobHooks::fulfill_hook(
+                &who,
+                &fulfillment,
+                requester.clone(),
+                &registration,
             )?;
 
-            // route fulfillment
+            // only route fulfillment after all checks succeeded since this call is not revertable
```
Comment (RomarQ) [inline pallets/acurast/src/lib.rs:334]: I am 99% sure this is not true.

Everything should get reverted, if the call returns `Err`.

File: pallets/marketplace/src/payments.rs [hunk 7]
```diff
@@ -21,6 +21,67 @@ impl<Asset> AssetBarrier<Asset> for () {
     }
 }
 
+pub(crate) type RewardFor<T> = <<T as Config>::RewardManager as RewardManager<T>>::Reward;
```
Comment (RomarQ) [inline pallets/marketplace/src/payments.rs:24]: This should probably be moved to `types.rs`

```suggestion
pub type RewardFor<T> = <<T as Config>::RewardManager as RewardManager<T>>::Reward;
```

File: pallets/marketplace/src/types.rs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+/// Represents an evaluation of the SLA after a job's schedule is completed.
+#[derive(RuntimeDebug, Encode, Decode, MaxEncodedLen, TypeInfo, Clone, PartialEq, Copy)]
+pub struct SLAEvaluation {
+    total: u8,
+    met: u8,
+}
+
+pub(crate) type JobRequirementsFor<T> = JobRequirements<RewardFor<T>>;
```
Comment (RomarQ) [inline pallets/marketplace/src/types.rs:68]:
```suggestion
pub type JobRequirementsFor<T> = JobRequirements<RewardFor<T>>;
```

File: pallets/marketplace/src/weights.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	fn delete_advertisement() -> Weight {
+		Weight::from_ref_time(22_000_000 as u64)
+			.saturating_add(T::DbWeight::get().reads(2 as u64))
+			.saturating_add(T::DbWeight::get().writes(3 as u64))
+	}
+	// fn update_job_matching() -> Weight {
+	// 	Weight::from_ref_time(21_000_000 as u64)
+	// 		.saturating_add(T::DbWeight::get().reads(2 as u64))
+	// 		.saturating_add(T::DbWeight::get().writes(1 as u64))
+	// }
```
Comment (RomarQ) [inline pallets/marketplace/src/weights.rs:46]:
```suggestion
```

Comment (RomarQ) [reply pallets/marketplace/src/weights.rs:46]: There is no `update_job_matching` call.

File: pallets/proxy/src/lib.rs [hunk 10]
```diff
@@ -98,7 +110,7 @@ pub mod pallet {
         // create an encoded version of the call
         let mut encoded_call = Vec::<u8>::new();
         // first byte is the pallet id on the destination chain
-        encoded_call.push(T::AcurastPalletId::get());
```
Comment (RomarQ) [inline pallets/proxy/src/lib.rs:101]: Why did this change?

Comment (RomarQ) [reply pallets/proxy/src/lib.rs:101]: We should use `T::AcurastPalletId::get()` preferably.

**Review by RomarQ — CHANGES_REQUESTED** [review]

File: pallets/marketplace/src/lib.rs (hunk 2 above)
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs]: RewardManager moved to marketplace

File: pallets/marketplace/src/lib.rs (hunk 3 above)
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs]: RewardManager moved to marketplace

File: pallets/marketplace/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+    pub type StoredCapacity<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, i32>;
+
+    /// Index with sorted advertisement by reward asset as a map [AssetId] -> Vec<([AccountId], [Price])>
+    #[pallet::storage]
+    #[pallet::getter(fn stored_ad_index)]
+    pub type StoredAdIndex<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        <T as Config>::AssetId,
+        Vec<AdvertismentIndexValue<T::AccountId, T::AssetAmount>>,
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs:94]: For the matching we will be always searching through a list of candidates. Keeping them in a sorted Vec allows for some optimisations in runtime at least.

File: pallets/marketplace/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_acurast::Config {
+        type Event: From<Event<Self>>
+            + IsType<<Self as pallet_acurast::Config>::Event>
+            + IsType<<Self as frame_system::Config>::Event>;
+        /// Extra structure to include in the registration of a job.
+        type RegistrationExtra: IsType<<Self as pallet_acurast::Config>::RegistrationExtra>
+            + Into<JobRequirements<RewardFor<Self>>>
+            + Parameter
+            + Member;
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs:48]: Removed, also dropped unneeded traits from other types in config

File: pallets/marketplace/README.md [hunk 13]
```diff
@@ -0,0 +1,400 @@
+# Acurast Pallet
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/README.md:1]: done

File: pallets/marketplace/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+[dev-dependencies]
+base64 = { version = "0.13.0", default-features = false, features = ["alloc"] }
+hex-literal = "0.3"
+
+sp-version = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.29" }
+sp-core = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.29" }
+pallet-balances = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.29" }
+parachain-info = { git = "https://github.com/paritytech/cumulus", default-features = false, branch = "polkadot-v0.9.29" }
+
+acurast-p256-crypto = { path = "../../p256-crypto" }
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/Cargo.toml:53]: removed

File: pallets/marketplace/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,69 @@
+[package]
+name = "pallet-acurast-marketplace"
+authors = ["Papers AG"]
+description = "FRAME pallet with the Acurast marketplace."
+version = "0.0.1"
+license = "Unlicense"
+homepage = "https://substrate.io"
+repository = "https://github.com/paritytech/substrate/"
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/Cargo.toml:8]: done

File: pallets/acurast/src/lib.rs (hunk 6 above)
Comment (gitsimon) [reply pallets/acurast/src/lib.rs:334]: removed comment

File: pallets/marketplace/src/payments.rs [hunk 16]
```diff
@@ -21,6 +21,67 @@ impl<Asset> AssetBarrier<Asset> for () {
...
+
+    fn lock_reward(
+        reward: Self::Reward,
+        owner: <T::Lookup as StaticLookup>::Source,
+    ) -> Result<(), DispatchError>;
+    fn pay_reward(
+        reward: Self::Reward,
+        target: <T::Lookup as StaticLookup>::Source,
+    ) -> Result<(), DispatchError>;
+}
```
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/src/payments.rs:65]: not sure since traits is currently not `pub`. I feel the separation should _by domain_, not by rust language construct, I'm open to discuss this.

File: pallets/proxy/src/lib.rs (hunk 10 above)
Comment (gitsimon) [reply pallets/proxy/src/lib.rs:101]: We are still using `T::AcurastPalletId::get()` and `T::AcurastMarketplacePalletId::get()` to choose where to send the call, but now we pass that over a newly introduced parameter `palletId` to this very function

Comment (RomarQ) [reply pallets/proxy/src/lib.rs:101]: In my opinion we should have a single getter that either gets `pallet_acurast` or `pallet_acurast_merketplace` id's, not both simultaneously.

I am fine ignoring this improvement on this MR, but we should look at it once the MR gets merged.

Things that can be done:

1. Having 2 proxy implementations, one for the standalone and other for the merketplace.
2. Add a feature to disable the advertisement logic.

File: pallets/marketplace/src/payments.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (RomarQ) [reply pallets/marketplace/src/payments.rs:65]: Traits are public. Have a look at file lib.rs. `pub use traits::*;`

File: pallets/marketplace/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (RomarQ) [reply pallets/marketplace/README.md:1]: It still appears for me.

File: pallets/marketplace/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (RomarQ) [reply pallets/marketplace/src/lib.rs:94]: Ok, we should do some benchmarks on this component. (For example: Adding 5000 processors and see the costs).

The current approach is susceptible to DOS attacks.

File: pallets/marketplace/src/weights.rs (hunk 9 above)
Comment (gitsimon) [reply pallets/marketplace/src/weights.rs]: removed

File: pallets/marketplace/src/lib.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (gitsimon) [reply pallets/marketplace/src/lib.rs]: you are right. in a separate take we will select who can advertise

File: pallets/marketplace/src/weights.rs (hunk 9 above)
Comment (gitsimon) [reply pallets/marketplace/src/weights.rs]: removed it

**Comment by RomarQ** [comment]
Also commit [temp: comment out failing tests](https://github.com/Acurast/acurast-core/pull/26/commits/ab996d5185bc7dae7fb9cb5560b14661dd6f2094) should be removed.

**Comment by RomarQ** [comment]
Some commits could be squashed.

**Review by RomarQ — APPROVED** [review]
